packages/flutter_tools/lib/src/compile.dart

```diff
@@ -752,24 +782,30 @@ class DefaultResidentCompiler implements ResidentCompiler {
   }
 
   @override
-  Future<CompilerOutput> compileExpression(
+  Future<ExpressionCompilerOutput> compileExpression(
     String expression,
     List<String> definitions,
     List<String> typeDefinitions,
     String libraryUri,
     String klass,
     bool isStatic,
-  ) {
+  ) async {
     if (!_controller.hasListener) {
       _controller.stream.listen(_handleCompilationRequest);
     }
 
     final Completer<CompilerOutput> completer = Completer<CompilerOutput>();
-    _controller.add(
-        _CompileExpressionRequest(
-            completer, expression, definitions, typeDefinitions, libraryUri, klass, isStatic)
-    );
-    return completer.future;
+    final _CompileExpressionRequest request =  _CompileExpressionRequest(
+        completer, expression, definitions, typeDefinitions, libraryUri, klass, isStatic);
+    _controller.add(request);
+    final CompilerOutput output = await completer.future;
+    if (output == null || output.errorCount > 0) {
+      request.fileRead.complete();
+      return ExpressionCompilerOutput(null);
+    }
+    final Uint8List bytes = _fileSystem.file(output.outputFilename).readAsBytesSync();
```

Comment:
Would it be simpler if CompilerOutput for compile expression request had dill file contents instead of a filename?
Would there still be a need for fileRead.complete/flush?

Reply:
In that case no, though I haven't thought through all of the implications. I think there are circumstances where the tool performs a full compilation without using the results, and I'd be worried about slowing that case down.

I think that only happens when we first start up the tool though.

Reply:
Ahh, so the issue is web compilation. We actually compile to 3 different files and use the output file name as a key to look that up. I could untangle that and make it a separate kind of compilation request - but it will substantially more refactoring.

Reply:
It looks like compile expression requests for vm and for the web are too different classes. We can only request bytes for vm compile expression request, no?
Here is a diff with mock-up of what I have in mind(haven't actually tried it) https://gist.github.com/aam/cb0eed24c58d0c0f107be9a89b9e93f3

Reply:
Ahh yeah, I think that work pretty well. I'll update this PR (well, tomorrow)

Reply:
updated